### PR TITLE
Add support for DependencyVersion to NuGet.exe cli UpdateCommand

### PR DESCRIPTION
**Pull request by ThomasSmeets**
Adding -DependencyVersion to NuGet Update

Linked to issue : https://github.com/NuGet/Home/issues/7694

Details: 
* Add new helper to handle the DependencyVersion, to avoid to duplicate code between the Install and Update command.
* Add the related unit tests to ensure the default behavior still works and the new features work as expect

## Testing/Validation

Tests Added: Yes
Reason for not adding tests:  
Validation:

### CODE REVIEW

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs
Comment (nkolev92) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs]: Use a Theory and providing the values there. 
Otherwise this provides a moving target if the tests ever fail. (you don't want a fail fast behavior here).

**Review by nkolev92 — COMMENTED** [review]
All caveats of packages.config and nuget.exe update/install considered, this looks fine :) 
Please change the test to use a theory instead.

File: src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs
Comment (zivkan) [inline src/NuGet.Clients/NuGet.CommandLine/Commands/InstallCommand.cs]: nitpick: formatting (whitespace after commas)

File: src/NuGet.Clients/NuGet.CommandLine/Common/DependencyBehaviorHelper.cs
Comment (zivkan) [inline src/NuGet.Clients/NuGet.CommandLine/Common/DependencyBehaviorHelper.cs]: nitpick: formatting (space after comma)

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs
Comment (zivkan) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs:1547]: nitpick: some parts of this method, particularly near the end, have consecutive empty lines.  It's not clear to me if there's a reason and no other tests in the file do it, so if there is a reason, it's better to use comments. Some tests have `// arrange`, `// act`, `// assert` comments to show the different phases of the test, in case that was the intention.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs
Comment (zivkan) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs]: nitpick: I think `a1Identity` would be a better name, especially given there's a `a1Package` below, but I see that other tests in the file use the same name patterns, so I understand if you were copying them.

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs
Comment (zivkan) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs]: nitpick: similarly `a2Identity` would be more clear

File: test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs
Comment (zivkan) [inline test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetUpdateCommandTests.cs]: nitpick: `result` might be a clearer name, but I know other tests just use `r`.

**Review by zivkan — APPROVED** [review]
First, thanks for the contribution. We appreciate you taking the time to help the product.

I didn't find any product issues, just some comments on formatting and naming (although the naming issues were copied from other tests, so it's great to keep the style of the existing code, it's just a shame the existing code could use improvement). I'll discuss with @nkolev92 if we want to wait for the formatting improvements or if we'll just merge it as is when we're ready.

File: src/NuGet.Clients/NuGet.CommandLine/Common/DependencyBehaviorHelper.cs
Comment (nkolev92) [reply src/NuGet.Clients/NuGet.CommandLine/Common/DependencyBehaviorHelper.cs]: Format the whole document as it's a new one :)

**Comment by ThomasSmeets** [comment]
Hello, 
I've changed the few formatting problems you have noticed.

File: src/NuGet.Clients/NuGet.CommandLine/Common/DependencyBehaviorHelper.cs
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.CommandLine/Common/DependencyBehaviorHelper.cs]: you can return here, and not even use the extra dependencyBehavior variable. 

Then at the bottom you just return defaultBehavior.

File: src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx:5416]: I wouldn't rename the name of the resource. 
It'd break localization without more changes.

**Review by nkolev92 — COMMENTED** [review]
A couple of small things.

File: src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx
Comment (ThomasSmeets) [reply src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx:5416]: I can switch back to  **InstallCommandUnknownDependencyVersion** but, wouldn't it be strange to have a resource name called "Installxxx" even if this resource is used for "Update" behavior.

Furthermore I didn't found any localization for this resource (usually the resource seems localized with " ResourceName_Language"  but i didn't find any of them for  "InstallCommandUnknownDependencyVersion".
Do the localizations are contained in another project?

Comment (nkolev92) [reply src/NuGet.Clients/NuGet.CommandLine/NuGetResources.resx:5416]: Nope, they are not, they usually in the same file. 
You're right this wouldn't affect localization so renaming it is ok. 
Nice catch :)

**Review by nkolev92 — APPROVED** [review]

**Comment by nkolev92** [comment]
fyi @karann-msft @aortiz-msft

File: src/NuGet.Clients/NuGet.CommandLine/Common/DependencyBehaviorHelper.cs
Comment (zivkan) [inline src/NuGet.Clients/NuGet.CommandLine/Common/DependencyBehaviorHelper.cs:4]: Please add the same license header that you see in all other files

**Review by zivkan — COMMENTED** [review]
I'm sorry we dropped the ball on this.

Can you please add the license header in the new files, rebase your changes onto the latest dev branch, and force push your branch? We can't merge it as it is now.

**Comment by ThomasSmeets** [comment]
> I'm sorry we dropped the ball on this.
> 
> Can you please add the license header in the new files, rebase your changes onto the latest dev branch, and force push your branch? We can't merge it as it is now.

I am facing lot of problems to proceed to the rebase. ( mostly I am unable to rebase without requiring to resolve all the conflicts during the past year , which will certainly lead to errors). Any tips to make it easier ?

**Comment by nkolev92** [comment]
@ThomasSmeets 
Looking at your commit maybe you merged instead of rebased? I see the delta is huge here. 

If you can go to the `before rebase` state, then I'd recommend squashing all your commits into 1. 
Maybe even someone on the team can you do it for you if you can reset back to your before rebase state. We have push permissions to forks.

**Comment by ThomasSmeets** [comment]
Had some problems with the rebase. 
But now it seems ok, you only have the a single commit with my wholes changes inside. I hope it will be fine.

**Comment by zivkan** [comment]
Unfortunately our CI logs aren't public, so I need to copy details here. It's failing everywhere because of a compile error:

> ##[error]src\NuGet.Clients\NuGet.CommandLine\Common\DependencyBehaviorHelper.cs(20,27): Error CS0246: The type or namespace name 'CommandLineException' could not be found (are you missing a using directive or an assembly reference?)

> ##[error]src\NuGet.Clients\NuGet.CommandLine\Commands\UpdateCommand.cs(288,89): Error CS0103: The name 'DependencyBehavior' does not exist in the current context

**Comment by zivkan** [comment]
@ThomasSmeets Thank you for your contribution. This should be available from NuGet 5.6 preview 2 and later.

**Pull request merged**
